include/display.h

```diff
@@ -0,0 +1,312 @@
+ * @brief Callback API for writing bitmap
+ * See display_write_bitmap() for argument description
+ */
+typedef int (*display_write_bitmap_api)(const struct device *dev, const u16_t x,
+					const u16_t y, const u16_t w,
+					const u16_t h, const u16_t p,
+					const void *data);
+
+/**
+ * @typedef display_read_bitmap_api
+ * @brief Callback API for reading bitmap
+ * See display_read_bitmap() for argument description
+ */
+typedef int (*display_read_bitmap_api)(const struct device *dev, const u16_t x,
+				       const u16_t y, const u16_t w,
+				       const u16_t h, const u16_t p,
+				       void *data);
+
+/**
+ * @typedef display_get_framebuffer_api
+ * @brief Callback API to get framebuffer pointer
+ * See display_get_framebuffer() for argument description
+ */
+typedef void *(*display_get_framebuffer_api)(const struct device *dev);
+
+/**
+ * @typedef display_set_brightness_api
+ * @brief Callback API to set display brightness
+ * See display_set_brightness() for argument description
+ */
+typedef int (*display_set_brightness_api)(const struct device *dev,
+					  const u8_t brightness);
+
+/**
+ * @typedef display_set_contrast_api
+ * @brief Callback API to set display contrast
+ * See display_set_contrast() for argument description
+ */
+typedef int (*display_set_contrast_api)(const struct device *dev,
+					const u8_t contrast);
+
+/**
+ * @typedef display_get_capabilities_api
+ * @brief Callback API to get display capabilities
+ * See display_get_capabilities() for argument description
+ */
+typedef void (*display_get_capabilities_api)(const struct device *dev,
+		struct display_capabilities *capabilities);
+
+/**
+ * @typedef display_set_pixel_format_api
+ * @brief Callback API to set pixel format used by the display
+ * See display_set_pixel_format() for argument description
+ */
+typedef int (*display_set_pixel_format_api)(const struct device *dev,
+		const enum display_pixel_format pixel_format);
+
+/**
+ * @brief Display driver API
+ * API which a display driver should expose
+ */
+struct display_driver_api {
+	display_on_api display_on;
+	display_off_api display_off;
+	display_write_bitmap_api write_bitmap;
+	display_read_bitmap_api read_bitmap;
+	display_get_framebuffer_api get_framebuffer;
+	display_set_brightness_api set_brightness;
+	display_set_contrast_api set_contrast;
+	display_get_capabilities_api get_capabilities;
+	display_set_pixel_format_api set_pixel_format;
+};
+
+/**
+ * @brief Write bitmap
+ *
+ * @param dev Pointer to device structure
+ * @param x x coordinate of the upper left corner
+ * @param y y coordinate of the upper left corner
+ * @param w width of the bitmap in pixels
+ * @param h height of the bitmap in pixels
+ * @param p pitch between lines in pixels
+ * @param data pointer to the data array, the data array should be at
+ * minimum p * h * bytes per pixel
+ *
+ * @retval 0 on success else negative errno code.
+ */
+inline int display_write_bitmap(const struct device *dev, const u16_t x,
```

Comment:
I do not agree with that. Generally, not every controller allows writing any size of data to any address. As generous as the API is currently it will make the drivers too complex. I would remove w, h and p (???) and define the region by x, y, the size of the buffer in octets and current_pixel_format. Ditto for the display_read_bitmap.

Reply:
Moving to X,Y only would not be an option as this would require to update the remainder of the line and this will significantly slow down the driver and increase the memory buffer requirements. For example with a 320x240 LCD writing a data of 10(w)x10(h) pixels on coordinates 0 (x),0(y) would require 3200 pixels instead of 100 pixels.

Most LCDs with integrated framebuffer for embedded applications support windowing and you can tell them to only to write to this window by setting x, y, w and h.

On the other hand some (monochrome) devices pack multiple pixels in a single byte either in horizontal or vertical direction and such devices can not accept freely all possible inputs.

Would it be acceptable to add 2 extra parameters (minW, minH) to the display capabilities structure and the caller of the API need to comply to them? So if we would assume we have a display with 8 pixel vertical packing and we would like to write 10x10 data we actually have to pass 16x10 data.

On the p (pitch parameter) see following comment:
https://github.com/zephyrproject-rtos/zephyr/pull/7833#discussion_r209240596

Reply:
>Would it be acceptable to add 2 extra parameters (minW, minH) to the display capabilities structure and the caller of the API need to comply to them? So if we would assume we have a display with 8 pixel vertical packing and we would like to write 10x10 data we actually have to pass 16x10 data.

I think it is sufficient to specify the direction, e.g. SCREEN_INFO_MONO_VTILED see https://github.com/zephyrproject-rtos/zephyr/pull/9035/commits/588e7b6c02eb8aad537aa9ebd1c43082d3e36a43

>On the p (pitch parameter) see following comment:
#7833 (comment)

I do not get it. I know pitch of as the number of octets for each row on the screen.

Reply:
> I think it is sufficient to specify the direction, e.g. SCREEN_INFO_MONO_VTILED see 588e7b6

lgtm, how do we proceed on this, move the code to this PR or keep it in the char framebuffer PR?

> I do not get it. I know pitch of as the number of octets for each row on the screen.

Buffer memory consumption could be reduced by performing some operations in place (resize, rotation) bit possibly at expense of more transfer operations

Reply:
>lgtm, how do we proceed on this, move the code to this PR or keep it in the char framebuffer PR?

It can stay as it is.

>Buffer memory consumption could be reduced by performing some operations in place (resize, rotation) bit possibly at expense of more transfer operations

I'm still not sure how it will be used. Can you give an example and insert it into the description of the parameter?

It would also be nice to have an API to rotate the image on the display by 180 °. Specifically, to change scanning start position if it is supported by the controller.

Reply:
> It would also be nice to have an API to rotate the image on the display by 180 °. Specifically, to change scanning start position if it is supported by the controller.

Could be useful to allow changing the screen orientation at runtime  but would it not be better the to allow increments of 90 degrees (if supported)?
E.g.
```
enum display_orientation {
  display_orientation_normal,
  display_orientation_flipped, or display_orientation_180_degrees, 
  display_orientation_clockwise_90_degrees,
  display_orientation_counter_clockwise_90_degrees,
};

int display_set_orientation(const struct device *dev, enum display_orientation orientation);
```